### DISCUSSION TITLE
Link isn't working

### DISCUSSION THREAD

**Discussion by ru4ert** [vercel/next.js#60675] [Help] [answered]
### Summary

I imported a link:
```
"use client";
import Link from "next/link";
```

and used it `<Link href="/hellp">hello</Link>;`

its not working on specific pages and i cannot track why. Did anyone faced the same issue? 

Are there side effects?

v14.0.4

if i hover the link i get this:
![image](https://github.com/vercel/next.js/assets/47078678/bce37b87-5c88-4cd5-bcc0-6652aba65f5b)

Inside the DOM
![image](https://github.com/vercel/next.js/assets/47078678/119270af-c8c2-43e3-8b6d-b59a77a8aa07)

The link isn't nested.

### Additional information

```js
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.2.0: Wed Nov 15 21:55:06 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T6020
Binaries:
  Node: 20.8.0
  npm: 10.1.0
  Yarn: 1.22.19
  pnpm: N/A
Relevant Packages:
  next: 14.0.4
  eslint-config-next: 14.0.4
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.0.4
Next.js Config:
  output: N/A
```

**Comment by nirvikpurkait** [comment]
Can you provide file and folder structure, and where you using the `Link` component?

**Reply by ru4ert** [reply]
Yes I can provide that:

```
[1.4K]  .
├── [1.1K]  app
│   ├── [ 160]  orderlists
│   │   ├── [ 128]  [id]
│   │   │   ├── [ 299]  layout.tsx
│   │   │   └── [ 40K]  page.tsx

```

Inside of the `orderlists/[id]/page.txt` i have the `link`


If i use a `<a>` the link works

**Reply by nirvikpurkait** [reply]
Check do you have typo of `/hellp` with **_ll_** insted of `/help` with **_l_**, when you have a folder named `help` with **_l_**

**Reply by nirvikpurkait** [reply]
Again you said that it works with `<a>`. 🤔🤔🤔

**Reply by ru4ert** [reply]
@nirvikpurkait 
i opened a bug: https://github.com/vercel/next.js/issues/60760

**Reply by icyJoseph** [reply, marked as answer]
@ru4ert the problem is that you have an infinite loop that keeps blocking navigation.

Using the `anchor` tag unloads the page, and barely manages to navigate. However, the infinite loop keeps React "stuck" and it cannot do the state update proper.

The infinite loop is here:

```jsx
   const { dispatchMobileNavState } = useContext(NavigationContext);
   useEffect(() => {
      if (dispatchMobileNavState) {
         dispatchMobileNavState({
            type: "set_float_menu",
            payload: <div>i am navi</div>
         })
      }
   })
```

This effect is what is called, a runaway effect, it batches an state update, which in turn, changes the context value, which in turn triggers the effect again, and so on.

You could place a `console.count('effect')` inside the effect, and before the App runs into an error, you'll see the count go up by a lot.

Anyhow, why does this effect loop forever? Because you are recreating the context on every state update:

```js
export const NavigationContext = createContext({
  mobileNavState: initialState,
  dispatchMobileNavState: (state, action) => { }
});

export const NavigationProvider = ({ children }) => {
  const [mobileNavState, dispatchMobileNavState] = useReducer(
    reducer,
    initialState,
  );

  return (
    <NavigationContext.Provider
      value={{ // this value object is a new object every time NavigationProvider renders
        mobileNavState,
        dispatchMobileNavState
      }}
    >
      {children}
    </NavigationContext.Provider>
  );
};
```

So, whenever, `NavigationProvider` renders, such as when `mobileNavState` changes, you recreate the `value` prop passed to `NavigationContext.Provider`, and that means that every component that consumes that context renders again. Which triggers the useEffect you have.

One quick way to remediate this is to make the effect, depend on the dispatcher:

```js
   useEffect(() => {
      if (dispatchMobileNavState) {
         dispatchMobileNavState({
            type: "set_float_menu",
            payload: <div>i am navi</div>
         })
      }
   }, [dispatchMobileNavState])
``` 

Because `dispatchMobileNavState`, is the dispatcher from useReducer, it'll be stable, and the effect runs once only. Ignoring ReactStrictMode for now.

In this case, the Card component does still render though. 

An even better solution, is to have two context providers, one for the state, and one for the dispatch:

```js
export const NavigationContext = createContext({...initialState});
const NavigationDispatch = createContext(null)

export const NavigationProvider = ({ children }) => {
  const [mobileNavState, dispatchMobileNavState] = useReducer(
    reducer,
    initialState,
  );

  return (
    <NavigationDispatch.Provider value={dispatchMobileNavState}>
      <NavigationContext.Provider value={mobileNavState}>
       {children}
      </NavigationContext.Provider>
    </NavigationDispatch.Provider>
  );
};

export const useNavigationDispatch = () => {
  const ctx = useContext(NavigationDispatch)

  if (!ctx) throw new Error("Using context outside provider")

  return ctx
}
```

Notice, how I don't export the Dispatch Context, I think you should also not export `NavigationContext`, instead, make a `useNavigationState` hook, similarly to `useNavigationDispatch`, do the null checks you need to do, and export that instead.

Now we have a very nice situation, updates to `mobileState` won't trigger re-render of components that only use the `navigation dispatch`, so you wouldn't, though maybe you should, need to use dependencies in the use-effect in Card.

Why it seems like it works when you mutate the state:

> To fix this you need to assign all values to the object, (not creating a new one, e.g with a spread operator):

Happens because you are mutating the state, and React sees the same object reference, so no renders are triggered, and so no infinite loop is created.

**Reply by ru4ert** [reply]
@icyJoseph thanks for now, i will track this issue in the original project

**Reply by ru4ert** [reply]
@ru4ert yep, that was the issue!
